move/utils/sources/utils/utils.move

```diff
@@ -0,0 +1,19 @@
+/// Module: utils
+module utils::utils;
+
+use sui::bcs::{Self, BCS};
+
+/// Remaining data after BCS decoding
+const ERemainingData: u64 = 101;
+
+public macro fun peel_data<$T>(
+    $data: vector<u8>,
+    $peel_fn: |&mut BCS| -> $T,
+): $T {
+    let mut bcs = bcs::new($data);
+    let result = $peel_fn(&mut bcs);
+    assert!(bcs.into_remainder_bytes().length() == 0, error_remaining_data());
+    result
+}
+
+public fun error_remaining_data(): u64 { ERemainingData }
```

Comment:
Why use a function for it?

mysten team recommend using this error pattern for more debugging info
```
#[error]
const ENotImplemented: vector<u8> = b"Function is not implemented";
```

Let's use that?

Reply:
The macro function will be executed the operation inline, from another module in our case, that means it cannot access to the internal variable (`ERemainingData`), so i need to wrap in inside the public function.

Reply:
ah, yes, macros are evaluated inline. Can you ask mysten team if this is possible somehow, or if they can add support in the future if not?

Reply:
@milapsheth As the Sui team has suggested, I've removed the error code from the `assert!` so that the error points at the macro invocation line. Additionally, to compensate the missing of error code, i've added more description about the error in the function docstring to give more details about potential error when debugging. I think this should be sufficient.